sklearn/manifold/_t_sne.py

```diff
@@ -745,6 +747,28 @@ class TSNE(BaseEstimator):
     (4, 2)
     """
 
+    _parameter_constraints = {
+        "n_components": [Interval(Integral, 1, None, closed="left")],
+        "perplexity": [Real],
+        "early_exaggeration": [Interval(Real, 1, None, closed="left")],
+        "learning_rate": [
+            StrOptions({"auto", "warn"}),
+            Interval(Real, 0, None, closed="neither"),
+        ],
```

Comment:
We don't want to advertise `"warn"` to the user, so we can use the `Hidden` constraint (from `utils._param_validation`)
```suggestion
        "learning_rate": [
            StrOptions({"auto"}),
            Hidden(StrOptions({"warn"})),
            Interval(Real, 0, None, closed="neither"),
        ],
```